### ISSUE TITLE
Support for tools / tool_choice in OpenAI-compatible API

### ISSUE THREAD

**Issue by TimPietrusky** [runpod-workers/worker-vllm#85]
vLLM doesn't support [`tools`](https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools) / `tool_choice` yet, [but hopefully at some point in time](https://github.com/vllm-project/vllm/pull/3237). This will make it possible to also use tools and build agents.  

Opening this to keep track of this feature and to inform anyone looking for this and wondering, why this is not working. I ran into these errors in the endpoint-logs on RunPod while testing tools:

### LangChain
```json
{
  "delayTime": 1135,
  "executionTime": 445,
  "id": "sync-123-123-123-123-123-123-123",
  "output": [
    {
      "code": 400,
      "message": "1 validation error for ChatCompletionRequest\nfunctions\n  Extra inputs are not permitted [type=extra_forbidden, input_value=[{'description': 'Get the...n'], 'type': 'object'}}], input_type=list]\n    For further information visit https://errors.pydantic.dev/2.7/v/extra_forbidden",
      "object": "error",
      "param": null,
      "type": "BadRequestError"
    }
  ],
  "status": "COMPLETED"
}
```

### Vercel AI SDK

```json
{
  "delayTime": 1106,
  "executionTime": 401,
  "id": "97c5be32-123-123-123-123-123",
  "output": [
    {
      "code": 400,
      "message": "2 validation errors for ChatCompletionRequest\ntool_choice\n  Extra inputs are not permitted [type=extra_forbidden, input_value='auto', input_type=str]\n    For further information visit https://errors.pydantic.dev/2.7/v/extra_forbidden\ntools\n  Extra inputs are not permitted [type=extra_forbidden, input_value=[{'function': {'descripti...}}, 'type': 'function'}], input_type=list]\n    For further information visit https://errors.pydantic.dev/2.7/v/extra_forbidden",
      "object": "error",
      "param": null,
      "type": "BadRequestError"
    }
  ],
  "status": "COMPLETED"
}
```

**Comment by Saran33** [comment]
vLLM now supports named function calling in the chat completion API for models that support it. However, the `tool_choice` options `auto` and `required` are not yet supported but on the roadmap.

To use a named function you need to explicitly call it in the `tool_choice` parameter. I haven't tried it yet personally.
- https://github.com/vllm-project/vllm/issues/1869
- https://vllm.readthedocs.io/_/downloads/en/stable/pdf/
- https://github.com/dongxiaolong/open-source-function-calling/blob/main/openhermes-functions-with-vllm.ipynb

There's an open PR to add `auto` tool choice from a set of tools.
- https://github.com/vllm-project/vllm/pull/5649

**Comment by TimPietrusky** [comment]
@Saran33 thanks for the info. 

I actually tried what you have suggested by explicitly calling my function, but this also leads to errors.

### TypeScript
When you set the name of the function as your `tool_choice`, you will get errors by TypeScript, as only `none`, `required` or `auto` are allowed. You can get around this by ignoring the error with a `// @ts-ignore`

### Vercel AI SDK
When the AI SDK tries to run the code, it will immidieatly complain, that the passed in function name is not an allowed option for `tool_choice`.  

You can't get around this error.

### LangChain

Here it is still called `function_call`, so when using that code, I get this error:

```json
{
  "delayTime": 1127,
  "executionTime": 138,
  "id": "sync-123-123-123-123-123-123-123",
  "output": [
    {
      "code": 400,
      "message": "2 validation errors for ChatCompletionRequest\nfunction_call\n  Extra inputs are not permitted [type=extra_forbidden, input_value={'name': 'get_current_weather'}, input_type=dict]\n    For further information visit https://errors.pydantic.dev/2.7/v/extra_forbidden\nfunctions\n  Extra inputs are not permitted [type=extra_forbidden, input_value=[{'description': 'Get the...n'], 'type': 'object'}}], input_type=list]\n    For further information visit https://errors.pydantic.dev/2.7/v/extra_forbidden",
      "object": "error",
      "param": null,
      "type": "BadRequestError"
    }
  ],
  "status": "COMPLETED"
}
```

**Comment by K-Mistele** [comment]
> @Saran33 thanks for the info.
> 
> I actually tried what you have suggested by explicitly calling my function, but this also leads to errors.
> 
> ### TypeScript
> When you set the name of the function as your `tool_choice`, you will get errors by TypeScript, as only `none`, `required` or `auto` are allowed. You can get around this by ignoring the error with a `// @ts-ignore`
> 
> ### Vercel AI SDK
> When the AI SDK tries to run the code, it will immediately complain, that the passed in function name is not an allowed option for `tool_choice`.
> 
> You can't get around this error.
> 
> ### LangChain
> ```json
> {
>   "delayTime": 1093,
>   "executionTime": 142,
>   "id": "sync-123-123-123-123-123-123",
>   "output": [
>     {
>       "code": 400,
>       "message": "2 validation errors for ChatCompletionRequest\ntool_choice\n  Extra inputs are not permitted [type=extra_forbidden, input_value='get_current_weather', input_type=str]\n    For further information visit https://errors.pydantic.dev/2.7/v/extra_forbidden\ntools\n  Extra inputs are not permitted [type=extra_forbidden, input_value=[{'description': 'Get the...n'], 'type': 'object'}}], input_type=list]\n    For further information visit https://errors.pydantic.dev/2.7/v/extra_forbidden",
>       "object": "error",
>       "param": null,
>       "type": "BadRequestError"
>     }
>   ],
>   "status": "COMPLETED"
> }
> ```

Are you running into this on the latest release of vLLM, or from a specific branch/PR?

**Comment by TimPietrusky** [comment]
@K-Mistele I'm using the image that comes from this repository: `runpod/worker-vllm:stable-cuda12.1.0`. According to the [commit of the submodule](https://github.com/runpod-workers/worker-vllm/commit/874379a0c5a59c400c9d9ad2d7db972a5c4bf5fc) this image is on vLLM 0.4.2

**Comment by Saran33** [comment]
@TimPietrusky ah yes, looking at the release there it was only [introduced in vLLM 0.5.0](https://github.com/vllm-project/vllm/releases/tag/v0.5.0). +1 on bumping the vLLM version to support this feature. To the best of my knowledge, other services like NVIDIA NIMS don't support tool calling either when testing their endpoints with the [example from the openAI docs](https://cookbook.openai.com/examples/how_to_call_functions_with_chat_models), so it would definitely make Runpod even more competitive. It looks like there are a few open issues related to bumping vLLM already https://github.com/runpod-workers/worker-vllm/issues/80 https://github.com/runpod-workers/worker-vllm/issues/83 https://github.com/runpod-workers/worker-vllm/issues/84

**Comment by TimPietrusky** [comment]
@Saran33 yeah, that would be great!

**Comment by eduardozamudio** [comment]
I've been trying to use [langgraph ChatOpenAI](https://python.langchain.com/v0.2/docs/integrations/chat/openai/) to allow bind tools on vLLM using OpenAI API (Meta-Llama-3-70B-Instruct). 

I've modified the `extra` argument on ConfigDict to **"allow"** in the OpenAIBAseModel class. As says [here](https://github.com/vllm-project/vllm/issues/4339#issuecomment-2076419315)

```python
# from vllm/entrypoints/openai/protocol.py

class OpenAIBaseModel(BaseModel):
    # OpenAI API does not allow extra fields
    model_config = ConfigDict(extra="allow") #Instead of extra="forbid"
```

With this modification, I've avoided the error of forbidden extra inputs and model responds. However, tools are never called.

**Comment by TimPietrusky** [comment]
@eduardozamudio thanks for trying this! And super sad, that it's not working :/

@alpayariyak what do you need to get this update going? How can we provide support?

**Comment by K-Mistele** [comment]
Yeah so auto tool choice doesn't work in vLLM yet - this is documented. I am working on adding it [here](https://github.com/vllm-project/vllm/pull/5649), but at this time tools will only work if you specify the tool to call with `tool_choice` since it can use outlines to force it. `tool_choice="auto"` and `tool_choice=None` are not supported - only `tool_choice="some_tool_name"` and `tool_choice="none"`

**Comment by TimPietrusky** [comment]
@K-Mistele awesome, THANK YOU VERY MUCH!!!

**Comment by pandyamarut** [comment]
@TimPietrusky- Really appreciate your help. You will hear super soon. Things have changed a bit, working on it.

**Comment by TimPietrusky** [comment]
@pandyamarut you are super welcome! I'm very excited to get this done, as we then can create AI agents on RunPod 😍

**Comment by pandyamarut** [comment]
@TimPietrusky  #82  is updated. I am currently doing sanity before it's released fully. In case, If you want to try it out. You can build an image and deploy a serverless worker, any issues & feedback you have, please let me know.

**Comment by TimPietrusky** [comment]
@pandyamarut thank you very much! Will test it tomorrow 🙏

**Comment by nerdylive123** [comment]
> @TimPietrusky #82 is updated. I am currently doing sanity before it's released fully. In case, If you want to try it out. You can build an image and deploy a serverless worker, any issues & feedback you have, please let me know.

Hey actually i tried building my serverless image for this vllm-worker, had some complications while building ( because its big, etc ), but got it pushed. and after i deployed it to a serverless endpoint i couldn't see the openai compatible url why is that? i also noticed its like for only specific tag from runpod/vllm-* (wildcard)

**Comment by pandyamarut** [comment]
Yeah that's specific to vllm-*. But you should still be able to use /v1/openai - to test. 

For faster updates there were some changes made which made image heavier.

**Comment by nerdylive123** [comment]
oh wow, okay thanks for the info~! didn't know that was possible

**Comment by TimPietrusky** [comment]
@K-Mistele thank you so much for putting so much time and energy in resolving all the things in your PR!!! We are so much looking forward to get this released in vLLM, you can't even imagine.

**Comment by K-Mistele** [comment]
Thanks 😁 Really looking forward to pushing this out

**Comment by TimPietrusky** [comment]
Thanks to @pandyamarut we have an updated vLLM to 0.5.3, which means we can actually use `tools` via https://github.com/runpod-workers/worker-vllm/releases/tag/v1.1. 

Will make sure to test this!

**Comment by TimPietruskyRunPod** [comment]
The PR was finally merged, looking forward to a new release.

Thank you @K-Mistele for putting SO MUCH WORK into this!!!

**Comment by K-Mistele** [comment]
@TimPietruskyRunPod no problem :) 
If you’d like I’m happy to keep you posted on future enhancements to this feature - I have already begun work on adding support for llama 3.1’s tool format

**Comment by TimPietrusky** [comment]
@K-Mistele oh yes please, that would be awesome!

**Comment by SmartManoj** [comment]
Is the issue solved?
